### ISSUE TITLE
[JS] Unable to build with webpack with strictExportPresence: true

### ISSUE THREAD

**Issue by tarekrached** [apache/arrow#2656]
This took me a while to track down, but the long and short of it is that the `apache-arrow` npm module can't be used in webpack with the `strictExportPresence` webpack config flag set to true.

This manifested when I tried using apache-arrow in a `create-react-app` application - it worked fine in the development mode, but failed when I tried to build the production bundle.  (Note that this was for `react-scripts@1.1.5` - in the nearly-completed `react-scripts@2`, dev mode fails, too.)

I have a [minimal reproduction of this issue here](https://github.com/tarekrached/apache-arrow-js-module-repro).

The following may be related:
- https://github.com/TypeStrong/ts-loader/issues/653
- https://github.com/Microsoft/TypeScript/issues/18008

**Comment by wesm** [comment]
@trxcllnt @TheNeuralBit seems like we should try to get a new JS release out there. Could the data refactor wait until 0.5.0?

**Comment by trxcllnt** [comment]
@wesm yes definitely, I think that was the plan all along. I've also run into this issue with webpack, so we should fix this before 0.4.0. I'm not entirely sure how this webpack flag works, since the TS does compile down to compliant syntax for CommonJS and ESModules:

`Arrow.js`:
```js
const table_1 = require("./table");
exports.Table = table_1.Table;
```

and the corresponding one in the ESModules file `Arrow.mjs`:
```js
import { Table, CountByResult } from './table';
/* ... */
export { Table, CountByResult };
```

@tarekrached do you have any ideas on what the fix would be?

**Comment by tarekrached** [comment]
@trxcllnt I don't have a clue what the fix would be, sorry.  I tried messing around with some of the ts compiler settings in `js/tsconfig` and rebuilding, but to no avail.  The TypeScript build process is something of a mystery to me.

FWIW it looks like this no longer manifests in the hot-off-the-presses `react-scripts@2.0.1`.  `npm start` and `npm run build` both work fine, and they both failed in `react-scripts@2.0.0`.  I'm not clear on what changed, as `strictExportPresence` is still set to `true`, so I'm not confident that it won't pop up again in future releases.

**Comment by tarekrached** [comment]
I think the relevant react-scripts change may be here, in which they treat `*.mjs` files specially. 

https://github.com/facebook/create-react-app/commit/3ae3cf3678725f86fe34234e59088a65a9458b8b#diff-d2bddbd3bfb7051f9381474c844674fa

I say this because the error that surfaced in `npm start` was:
```
./node_modules/@apache-arrow/es2015-esm/type.js 27:18-29
  'flatbuffers' does not contain an export named 'flatbuffers'.
```
and the flatbuffers source was `node_modules/flatbuffers/js/flatbuffers.mjs`

**Comment by mlyknown** [comment]
everybody has some solution?

**Comment by tarekrached** [comment]
I spoke too soon - `@apache-arrow/es2015-esm` compiles, but `flatbuffers` is undefined:

![image](https://user-images.githubusercontent.com/167218/46320889-50ba8480-c595-11e8-936f-a7f6cdc1b073.png)

This is with `react-scripts` `2.0.1` and `2.0.2`

**Comment by tarekrached** [comment]
I think [this issue](https://github.com/jaydenseric/apollo-upload-client/issues/72) in `apollo-upload-client` is related.  AFAICT the issue stems from a named export in an upstream non-esm module.

also possibly related: https://github.com/webpack/webpack/issues/7973 Broken cjs & esm interop

**Comment by trxcllnt** [comment]
I believe the original issue in this thread has been resolved by create-react-app updates. There's still a ][separate Webpack 4 bug](https://github.com/webpack/webpack/issues/8491) that causes webpack to incorrectly compile when a project has CommonJS and ESM side-by-side. Apparently it has been fixed in Webpack 5, but `create-react-app` is still on Webpack 4 [until Webpack 5 is out of alpha](https://github.com/facebook/create-react-app/issues/6081#issuecomment-449500730).

There is a workaround for Webpack 4 that I've tested works, so I've [filed an report with CRA](https://github.com/facebook/create-react-app/issues/6289) about rolling back to an earlier config that worked in 4. In the meantime, I've been using the `@apache-arrow/esnext-cjs` build in Graphistry. It may be worth switching the main `apache-arrow` module to be one of the CommonJS builds for this next release, then following up when these external issues are resolved.
